src/Microsoft.JSInterop/DotNetDispatcher.cs

```diff
@@ -0,0 +1,181 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Threading.Tasks;
+
+namespace Microsoft.JSInterop
+{
+    /// <summary>
+    /// Provides methods that receive incoming calls from JS to .NET.
+    /// </summary>
+    public static class DotNetDispatcher
+    {
+        private static ConcurrentDictionary<string, IReadOnlyDictionary<string, (MethodInfo, Type[])>> _cachedMethodsByAssembly
+            = new ConcurrentDictionary<string, IReadOnlyDictionary<string, (MethodInfo, Type[])>>();
+
+        /// <summary>
+        /// Receives a call from JS to .NET, locating and invoking the specified method.
+        /// </summary>
+        /// <param name="assemblyName">The assembly containing the method to be invoked.</param>
+        /// <param name="methodIdentifier">The identifier of the method to be invoked. The method must be annotated with a <see cref="JSInvokableAttribute"/> matching this identifier string.</param>
+        /// <param name="argsJson">A JSON representation of the parameters.</param>
+        /// <returns>A JSON representation of the return value, or null.</returns>
+        public static string Invoke(string assemblyName, string methodIdentifier, string argsJson)
```

Comment:
Good point. Added the following comment to both:

```
// This method doesn't need [JSInvokable] because the platform is responsible for having
// some way to dispatch calls here. The logic inside here is the thing that checks whether
// the targeted method has [JSInvokable]. It is not itself subject to that restriction,
// because there would be nobody to police that. This method *is* the police.
```